pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,227 @@
+
+	cbBuilder := circuitbreaker.Builder[any]().
+		WithFailureThreshold(cfg.FailureThreshold).
+		WithDelay(cfg.CooldownPeriod).
+		OnFailure(func(failsafe.ExecutionEvent[any]) {
+			countError.Inc()
+		}).
+		OnSuccess(func(failsafe.ExecutionEvent[any]) {
+			countSuccess.Inc()
+		}).
+		OnClose(func(event circuitbreaker.StateChangedEvent) {
+			transitionClosed.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is closed", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionHalfOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is half-open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		HandleIf(func(_ any, err error) bool { return isFailure(err) })
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThreshold)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThreshold, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb := cbBuilder.Build()
+	return &circuitBreaker{
+		CircuitBreaker: cb,
+		ingester:       ingester,
+		executor:       failsafe.NewExecutor[any](cb),
+		startTime:      time.Now().Add(cfg.InitialDelay),
+	}
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	return cb.startTime.Before(time.Now())
+}
+
+func (cb *circuitBreaker) ingesterID() string {
+	return cb.ingester.cfg.IngesterRing.InstanceID
+}
+
+func (cb *circuitBreaker) logger() log.Logger {
+	return cb.ingester.logger
+}
+
+func (cb *circuitBreaker) metrics() *ingesterMetrics {
+	return cb.ingester.metrics
+}
+
+func (cb *circuitBreaker) config() CircuitBreakerConfig {
+	return cb.ingester.cfg.CircuitBreakerConfig
+}
+
+func (cb *circuitBreaker) get(ctx context.Context, callbackFnName string, callbackFn func() (any, error)) (any, error) {
```

Comment:
suggestion: rename "get" => "execute" rename `callback` to `operation`, ie.
* `callbackFnName` -> `operationName`
* `callbackFn` -> `op`

Reply:
I have renamed `callbackFnName` and `callbackFn` into `operationName` and `op` respectively, but I kept `get` because this method corresponds to the `failsafe.Executor.Get()` method, and it is used when the `Executor` needs to return a result (of type `any`).
Similarly, when we add the read path-related methods, there will be a `run()` method, which will call `failsafe.Executor.Run()`, which is used when no result is returned.

Reply:
After removing logs, `operationName` is actually not needed.